otherlibs/systhreads/st_stubs.c

```diff
@@ -15,6 +15,12 @@
 
 #define CAML_INTERNALS
 
+#if defined(_WIN32) && !defined(NATIVE_CODE)
+/* Ensure that pthread.h marks symbols __declspec(dllimport) so that they can be
+   picked up from the runtime (which will have linked winpthreads statically) */
+#define DLL_EXPORT
+#endif
```

Comment:
@Octachron and I don't quite understand what this is doing.  Which pthread.h file is being affected by this `#define`?

Reply:
`pthread.h` from the winpthreads library in mingw-w64 (`/usr/x86_64-w64-mingw32/sys-root/mingw/include/ptjhread.h`). If `DLL_EXPORT` is defined and `IN_WINPTHREAD` is not defined, then it decorates all the API functions with `__declspec(dllimport)` (as we would if they were declared `CAMLextern` in our headers).

That doesn't appear to be expressly documented - if for some reason the header changed that, the _possible_ impact would be the delightful "Cannot relocate" errors.

It has to be done here because `caml/domain.h` `#include`'s `caml/platform.h` which #include's `<pthread.h>` (i.e. by the time `st_pthreads.h` does `#include <pthread.h>` it has already happened).

Reply:
Thanks for the explanations.  We were looking at the wrong pthread.h file.

This trick is kind of fragile indeed, but no need to think of alternatives until it breaks.